src/generator/generator_visitor.ts

```diff
@@ -100,42 +103,81 @@ class GeneratorVisitor extends Visitor {
     protected visitBinaryExpression(node: BinaryExpression) {
         super.visitBinaryExpression(node);
 
-        const right = this.popExpression();
-        const left = this.popExpression();
+        let right = this.popExpression();
+        let left = this.popExpression();
+
+        const rightType = this.getExpressionType(node.right);
+        const leftType = this.getExpressionType(node.left);
+        const commonNumberType = getCommonNumberType(leftType, rightType);
+
+        left = this.convertType(left, leftType, commonNumberType);
+        right = this.convertType(right, rightType, commonNumberType);
+
+        const operationsInstance = this.getOperationsInstance(commonNumberType);
 
         switch (node.operator) {
             case '+':
-                this.expressions.push(this.module.i32.add(left, right));
+                this.expressions.push(operationsInstance.add(left, right));
                 break;
             case '-':
-                this.expressions.push(this.module.i32.sub(left, right));
+                this.expressions.push(operationsInstance.sub(left, right));
                 break;
             case '*':
-                this.expressions.push(this.module.i32.mul(left, right));
+                this.expressions.push(operationsInstance.mul(left, right));
                 break;
             case '/':
-                this.expressions.push(this.module.i32.div_s(left, right));
+                if (commonNumberType === WebAssemblyType.INT_32) {
+                    this.expressions.push(this.module.i32.div_s(left, right));
+                } else {
```

Comment:
Replace with an else if and throw an error in case it's neither `INT_32` nor `FLOAT_64`.